### PR TITLE
Feature/activities slug

### LINKED ISSUE #76: activities: get project/event by slug name
This is the same feature currently being done for posts (see #75 ). It would be much better to display `ni.fe.up.pt/projects/nijobs` instead of the ugly ID `ni.fe.up.pt/projects/2`, so we should store their unique slug names (customizable by the user) and be able to retrieve the activities using them.

Note: We decided it's best to leave the automatic `title -> slug` conversion to the frontend since titles and names might not be unique and the user can correct the slug in the form.

### PR DESCRIPTION
**Pull request by jamcunha**
Closes #76 

Similar to #75, I added an optional slug to the activities.

# Review checklist
-   [ ] Properly documents API changes in `docs/openapi.yml`
-   [x] Contains enough appropriate tests
-   [x] Behavior is as expected
-   [x] Clean, well structured code

### CODE REVIEW

File: src/main/kotlin/pt/up/fe/ni/website/backend/service/ProjectService.kt [hunk 1]
```diff
@@ -12,15 +12,27 @@ class ProjectService(private val repository: ProjectRepository) {
...
 
+    fun getProjectBySlug(projectSlug: String): Project = repository.findBySlug(projectSlug)
+        ?: throw NoSuchElementException(ErrorMessages.projectNotFound(projectSlug))
+
     fun updateProjectById(id: Long, dto: ProjectDto): Project {
         val project = getProjectById(id)
+
+        repository.findBySlug(dto.slug)?.let {
+            if (it.id != project.id) throw IllegalArgumentException(ErrorMessages.slugAlreadyExists)
+        }
```
Comment (BrunoRosendo) [inline src/main/kotlin/pt/up/fe/ni/website/backend/service/ProjectService.kt:43]: #55 needs to take this into consideration for the events

File: src/test/kotlin/pt/up/fe/ni/website/backend/controller/EventControllerTest.kt [hunk 2]
```diff
@@ -101,11 +137,13 @@ internal class EventControllerTest @Autowired constructor(
                 requiredFields = mapOf(
                     "title" to testEvent.title,
                     "description" to testEvent.description,
-                    "date" to testEvent.date
+                    "date" to testEvent.date,
+                    "slug" to testEvent.slug
```
Comment (BrunoRosendo) [inline src/test/kotlin/pt/up/fe/ni/website/backend/controller/EventControllerTest.kt:342]: Do we need this? Slug is not required

**Review by BrunoRosendo — CHANGES_REQUESTED** [review]
Good job! But aren't we missing the getBySlug endpoint for events?

**Review by BrunoRosendo — APPROVED** [review]
Looks good! You just need to rebase it with latest changes in develop

**Review by BrunoRosendo — APPROVED** [review]

**Review by bdmendes — COMMENTED** [review]
Good job! I have a small suggestion. Since both projects and events services and repositories are accessing the slug in the exact same way, what do you think of introducing an abstract `ActivityRepository` (and service)?

**Comment by MRita443** [comment]
> Good job! I have a small suggestion. Since both projects and events services and repositories are accessing the slug in the exact same way, what do you think of introducing an abstract `ActivityRepository` (and service)?

I implemented a similar solution in #89, check it out to see if it's what you suggested!

**Comment by bdmendes** [comment]
> > Good job! I have a small suggestion. Since both projects and events services and repositories are accessing the slug in the exact same way, what do you think of introducing an abstract `ActivityRepository` (and service)?
> 
> I implemented a similar solution in #89, check it out to see if it's what you suggested!

I like it! I would just make it a bit more cohesive by using inheritance.

**Comment by BrunoRosendo** [comment]
> > > Good job! I have a small suggestion. Since both projects and events services and repositories are accessing the slug in the exact same way, what do you think of introducing an abstract `ActivityRepository` (and service)?
> > 
> > 
> > I implemented a similar solution in #89, check it out to see if it's what you suggested!
> 
> I like it! I would just make it a bit more cohesive by using inheritance.

Maybe we should wait for #89 merge then. Inheritance might be trickier but it'd be nice to look at an implementation of that

**Review by bdmendes — COMMENTED** [review]
Looks good to me. However, I thought that the slug was intended to be a small internal representation of the title (perhaps with some parts of the date and/or description) and should be generated automatically upon creating the activity. Would not that be a better approach than the optional one?

**Comment by jamcunha** [comment]
> Looks good to me. However, I thought that the slug was intended to be a small internal representation of the title (perhaps with some parts of the date and/or description) and should be generated automatically upon creating the activity. Would not that be a better approach than the optional one?

We agreed in #75 that the slug would be generated in the frontend, giving the option to write another slug if wanted.

**Comment by bdmendes** [comment]
Awesome, let's get this merged then

**Review by bdmendes — APPROVED** [review]

**Pull request merged**